event_handlers.go

```diff
@@ -0,0 +1,100 @@
+package lightstep
+
+import (
+	"log"
+	"sync"
+	"sync/atomic"
+)
+
+func init() {
+	OnEvent(NewOnEventLogOneError())
+}
+
+var eventHandler atomic.Value
+var eventHandlerLock sync.Mutex
+
+// An OnEventHandler can be registered with OnEvent to
+type OnEventHandler func(Event)
+
+// onEvent is a thread-safe function for emiting tracer events.
+func onEvent(event Event) {
+	handler := eventHandler.Load().(OnEventHandler)
+	handler(event)
+}
+
+// OnEvent sets a global handler to receive tracer events as they occur. Events
+// may be emitted by the tracer, or by calls to static functions in this package.
+// It is suggested that you set your OnEvent handler before starting your tracer,
+// but it is safe to set a new handler at any time. Setting a new handler removes
+// the previous one.
+//
+// OnEvent is called synchronously – do not block in your handler as it may interfere
+// with the tracer. If no OnEvent handler is provided, a LogOnceOnError handler
+// is used by default.
+//
+// NOTE: Event handling is for reporting purposes only. It is not intended as a
+// mechanism for controling or restarting the tracer. Connection issues, retry
+// logic, and other transient errors are handled internally by the tracer.
+func OnEvent(handler OnEventHandler) {
+	eventHandlerLock.Lock()
```

Comment:
This lock isn't necessary, right? `atomic.Value` should be thread-safe.

Reply:
This needs more investigation in general, I'm still looking into the pitfalls of `atomic.Value`.

Reply:
Looks like no lock is needed here. Fun fact: using a typed function type rather than an interface type avoids a panic in atomic.Value.

```go
// this works well with atomic.Value
type EventHandler func(Event)

// using an interface instead causes atomic.Value to panic when you
// store a new value with a different underlying concrete type.
type EventHandler interface {
  OnEvent(Event)
}

// sticking the EventHandler interface in a stable concrete type 
// avoids that issue. 
type eventHandlerContainer struct {
  handler EventHandler
}
```

^ That was a little weird to me... when embedded, the interface object is treated as memory layout. On it's own, it's treated as language construct and the pointer to the concrete type is walked.